### PR TITLE
`get_ref` should never be called for an instance of reference count of zero

### PR DESCRIPTION
**Pull request by phoad**
https://github.com/rsocket/rsocket-cpp/issues/671

This will prevent buggy code to happen.

Next, we should also properly implement make_ref to let calling get_ref to be possible at the inside of the constructor.

### CODE REVIEW

File: yarpl/include/yarpl/Refcounted.h [hunk 1]
```diff
@@ -273,13 +273,26 @@ Reference<T> make_ref(Args&&... args) {
   return Reference<T>(new T(std::forward<Args>(args)...));
 }
 
-template <typename T>
+template <
+    typename T,
+    typename =
+    std::enable_if_t<std::is_base_of<Refcounted, std::decay_t<T>>::value>>
 Reference<T> get_ref(T& object) {
+  assert(
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: `DCHECK_GT(object.count(), 0) << "get_ref should not...";`

File: yarpl/include/yarpl/Refcounted.h [hunk 2]
```diff
@@ -273,13 +273,26 @@ Reference<T> make_ref(Args&&... args) {
   return Reference<T>(new T(std::forward<Args>(args)...));
 }
 
-template <typename T>
+template <
+    typename T,
+    typename =
+    std::enable_if_t<std::is_base_of<Refcounted, std::decay_t<T>>::value>>
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: I'd do this as a `static_assert()` inside the method personally.

File: yarpl/include/yarpl/Refcounted.h (hunk 1 above)
Comment (phoad) [inline yarpl/include/yarpl/Refcounted.h]: DCHECK_GT depends on including gtest related header. I did not want to include that in this header as this is a low level header.
We all use `assert` in this file.

File: yarpl/include/yarpl/Refcounted.h (hunk 2 above)
Comment (phoad) [inline yarpl/include/yarpl/Refcounted.h]: Why static_assert ?

Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: Cuts down on unnecessary template params.  They can get quite verbose when looking at stack traces 😄 .

File: yarpl/include/yarpl/Refcounted.h (hunk 1 above)
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: Technically it's glog and not gtest, so it's not as heavyweight.

I generally dislike using `assert()` as it tends to break opt builds with that have `-Wno-unused-variable` set.

File: yarpl/include/yarpl/Refcounted.h (hunk 2 above)
Comment (phoad) [inline yarpl/include/yarpl/Refcounted.h]: Oo, you would like to also prevent any overloaded one to have the error too.

**Pull request merged**